applications/ShallowWaterApplication/custom_processes/depth_integration_process.cpp

```diff
@@ -146,55 +147,78 @@ void DepthIntegrationProcess::Integrate(std::vector<PointerVector<GeometricalObj
-        if (!mVelocityDepthIntegration) {
+        const auto config = Globals::Configuration::Current;
+        Vector shape_functions;
+        BruteForcePointLocator locator(mrVolumeModelPart);
+
+        if (mVelocityDepthIntegration)
+        {
+            // Integrate the velocity over the water column
+            const int num_steps = 20;
+            const double step = (height -0.2 * rObjects.begin()->GetGeometry().Length()) / double(num_steps-1);
+            const double weight = 1.0 / double(num_steps);
+            const array_1d<double,3> start = rNode + mDirection * (min_elevation -inner_prod(rNode, mDirection));
+            Point point(start);
+
+            auto elem_id = locator.FindObject(rObjects, point, shape_functions, config);
+            array_1d<double,3> point_vel = ZeroVector(3);
+            if (elem_id > 0) {velocity += 0.5 * weight * InterpolateVelocity(elem_id, shape_functions);
+            }
+            array_1d<double,3> obj_velocity;
+            for (int i = 1; i < num_steps; ++i) {
+                point += step * mDirection;
+                elem_id = locator.FindObject(rObjects, point, shape_functions, config);
+                if (elem_id > 0) {
+                    obj_velocity = InterpolateVelocity(elem_id, shape_functions);
+                    velocity += weight * obj_velocity;
+                }
+            }
+            velocity -= 0.5 * weight * obj_velocity;
+        }
+        else
+        {
+            // Evaluate the velocity at a certain depth
             const double reference_depth = mMeanWaterLevel - min_elevation;
             const double target_depth = mMeanWaterLevel + mVelocityRelativeDepth * reference_depth;
             const double target_distance = target_depth - inner_prod(mDirection, rNode);
-            array_1d<double,3> target_point = rNode + mDirection * target_distance;
-            Point local_coordinates;
-            velocity = ZeroVector(3);
-            for (auto& object : rObjects) {
-                if (object.GetGeometry().IsInside(target_point, local_coordinates)) {
-                    for (auto& r_node : object.GetGeometry()) {
-                        velocity += r_node.FastGetSolutionStepValue(VELOCITY);
-                    }
-                    velocity /= object.GetGeometry().size();
-                }
+            const Point target_point(rNode + mDirection * target_distance);
+            auto elem_id = locator.FindObject(rObjects, target_point, shape_functions, config);
+            if (elem_id > 0) {
+                velocity = InterpolateVelocity(elem_id, shape_functions);
             }
         }
     }
+    array_1d<double,3> momentum = height * velocity;
     SetValue(rNode, MOMENTUM, momentum);
     SetValue(rNode, VELOCITY, velocity);
     SetValue(rNode, HEIGHT, height);
 }
 
+array_1d<double,3> DepthIntegrationProcess::InterpolateVelocity(const int ElementId, const Vector& rShapeFunctionValues) const
```

Comment:
```suggestion
array_1d<double,3> DepthIntegrationProcess::InterpolateVelocity(
    const int ElementId,
    const Vector& rShapeFunctionValues) const
```
Minor style guide suggestion.